File: TracksDemo/TracksDemo/ViewController.m
#import "ViewController.h"
@import AutomatticTracks;

@interface ViewController () <NSFetchedResultsControllerDelegate>

@property (nonatomic, strong) TracksContextManager *contextManager;
@property (nonatomic, strong) TracksService *tracksService;
@property (nonatomic, strong) NSFetchedResultsController *fetchedResultsController;

@property (nonatomic, weak) IBOutlet UIProgressView *progressView;
@property (nonatomic, weak) IBOutlet UILabel *objectCountLabel;
@property (nonatomic, strong) NSTimer *timer;
@property (nonatomic, strong) NSDate *startTime;

@property (nonatomic, weak) IBOutlet UISwitch *remoteCallsSwitch;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.contextManager = [TracksContextManager new];
    self.tracksService = [[TracksService alloc] initWithContextManager:self.contextManager];
    self.tracksService.queueSendInterval = 10.0;
    [self resetTimer];
    
    
    NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"TracksEvent"];
    NSSortDescriptor *sortDescriptor = [NSSortDescriptor sortDescriptorWithKey:@"date" ascending:YES];
    fetchRequest.sortDescriptors = @[sortDescriptor];
    self.fetchedResultsController = [[NSFetchedResultsController alloc] initWithFetchRequest:fetchRequest
                                                                        managedObjectContext:self.contextManager.managedObjectContext
                                                                          sectionNameKeyPath:nil
                                                                                   cacheName:nil];
    self.fetchedResultsController.delegate = self;
    [self.fetchedResultsController performFetch:nil];
    [self updateObjectCountLabel];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(resetTimer) name:TrackServiceWillSendQueuedEventsNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(resetTimer) name:TrackServiceDidSendQueuedEventsNotification object:nil];
    
    [self switchToAnonymousUser];
}


- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}


#pragma mark - IBAction methods

- (IBAction)sendTestEvent:(id)sender
{
    [self.tracksService trackEventName:@"test_event"];
}


- (IBAction)sendTestEventWithProperties:(id)sender
{
    [self.tracksService trackEventName:@"test_event_with_properties" withCustomProperties:@{@"custom_prop_1" : @"valuetew"}];
}


- (IBAction)crashApplicationTapped:(id)sender
{
    abort();
}


- (IBAction)userTypeSegmentedControlChanged:(UISegmentedControl *)sender
{
    if (sender.selectedSegmentIndex == 0) {
        [self switchToAnonymousUser];
    } else {
        [self switchToWordPressComUser];
    }
}


- (IBAction)remoteCallsSwitchTapped:(UISwitch *)sender
{
    self.tracksService.remoteCallsEnabled = sender.isOn;
    
    if (sender.isOn) {
        [self resetTimer];
    } else {
        [self.timer invalidate];
    }
}


#pragma mark - Fetched results delegate methods

- (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)anObject atIndexPath:(NSIndexPath *)indexPath forChangeType:(NSFetchedResultsChangeType)type newIndexPath:(NSIndexPath *)newIndexPath;
{
    [self updateObjectCountLabel];
}


#pragma mark - Private helper methods

- (void)updateObjectCountLabel
{
    self.objectCountLabel.text = [NSString stringWithFormat:@"Number of events queued: %@", @(self.fetchedResultsController.fetchedObjects.count)];
}

- (void)resetTimer
{
    [self.timer invalidate];
    self.startTime = [NSDate date];
    self.progressView.progress = 0.0f;
    
    self.timer = [NSTimer scheduledTimerWithTimeInterval:self.tracksService.queueSendInterval / 100
                                                  target:self
                                                selector:@selector(timerFireMethod:)
                                                userInfo:nil
                                                 repeats:YES];
}


- (void)timerFireMethod:(NSTimer *)timer
{
    NSDate *fireDate = timer.fireDate;
    
    if ([fireDate timeIntervalSinceDate:self.startTime] > self.tracksService.queueSendInterval) {
        [self.timer invalidate];
    }
    
    CGFloat progress = [fireDate timeIntervalSinceDate:self.startTime] / self.tracksService.queueSendInterval;
    self.progressView.progress = progress;
}


- (void)switchToAnonymousUser
{
    [self.tracksService switchToAnonymousUserWithAnonymousID:[[NSUUID UUID] UUIDString]];
}


- (void)switchToWordPressComUser
{
    [self.tracksService switchToAuthenticatedUserWithUsername:@"astralbodies" userID:@"67137" skipAliasEventCreation:NO];
}

@end

File: TracksDemo/TracksDemo Mac/ViewController.swift
//
//  ViewController.swift
//  TracksDemo Mac
//
//  Created by Jeremy Massel on 2019-01-04.
//  Copyright © 2019 Automattic Inc. All rights reserved.
//

import Cocoa
import AutomatticTracks

class ViewController: NSViewController {

    private lazy var contextManager = TracksContextManager()
    private lazy var tracksService = TracksService(contextManager: contextManager)
    private var fetchedResultsController: NSFetchedResultsController<TracksEventCoreData>!

    @IBOutlet var progressView: NSProgressIndicator!
    @IBOutlet weak var queuedEventsLabel: NSTextField!
    
    var sendInterval: TimeInterval {
        get{ return tracksService?.queueSendInterval ?? 10 }
        set { tracksService?.queueSendInterval = newValue}
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        tracksService?.queueSendInterval = 10.0
        tracksService?.eventNamePrefix = "tracks_macos_demo"

        resetTimer()
        setupFetchedResultsController()
        addTimerEventListeners()
        switchToAnonymousUser()
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    //MARK: - IB Action Methods
    @IBAction func didClickTrackEvent(_ sender: NSButton) {
        tracksService?.trackEventName("test_event")
    }

    @IBAction func didClickTrackEventWithCustomProperties(_ sender: NSButton) {
        tracksService?.trackEventName("test_event_with_properties", withCustomProperties: [
            "custom_prop_1": "valuetew"
        ])
    }

    @IBAction func didClickCrashApplication(_ sender: NSButton ) {
        abort()
    }

    @IBAction func userTypeSegmentedControlDidChange(_ sender: NSSegmentedControl) {
        if sender.selectedSegment == 0 {
            switchToAnonymousUser()
        }
        else {
            switchToWordPressDotComUser()
        }
    }

    @IBAction func automaticallySendEventsToggled(_ sender: NSButton) {
        tracksService?.remoteCallsEnabled = sender.state == .on

        if sender.state == .on {
            resetTimer()
        }
        else{
            timer.invalidate()
        }
    }

    //MARK: – Fetched Results Controller
    private func setupFetchedResultsController(){
        let fetchRequest = NSFetchRequest<TracksEventCoreData>(entityName: "TracksEvent")
        fetchRequest.sortDescriptors = [ NSSortDescriptor(key: "date", ascending: true) ]

        self.fetchedResultsController = NSFetchedResultsController(
            fetchRequest: fetchRequest,
            managedObjectContext: contextManager.managedObjectContext,
            sectionNameKeyPath: nil,
            cacheName: nil)

        self.fetchedResultsController.delegate = self

        do{
            try self.fetchedResultsController.performFetch()
        }
        catch let err {
            debugPrint("Error fetching Tracks Events: \(err.localizedDescription)")
        }
    }

    //MARK: - Timer
    private lazy var timer = Timer()
    private var startTime: Date!

    private func addTimerEventListeners() {
        NotificationCenter.default.addObserver(self, selector: #selector(resetTimer), name: NSNotification.Name.TrackServiceWillSendQueuedEvents, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(resetTimer), name: NSNotification.Name.TrackServiceDidSendQueuedEvents, object: nil)
    }

    @objc private func resetTimer() {
        timer.invalidate()
        startTime = Date()
        progressView.doubleValue = 0

        timer = Timer.scheduledTimer(timeInterval: sendInterval / 100, target: self, selector: #selector(fireTimer(_:)), userInfo: nil, repeats: true)
    }

    @objc func fireTimer(_ sender: Timer) {
        if sender.fireDate.timeIntervalSince(startTime) > sendInterval {
            timer.invalidate()
        }

        let progress = sender.fireDate.timeIntervalSince(startTime) / sendInterval
        progressView.doubleValue = progress * 100
    }

    //MARK: - Helpers
    private func updateObjectCountLabel() {
        let count = fetchedResultsController.fetchedObjects?.count ?? 0
        queuedEventsLabel.stringValue = "Number of events queued: \(count)"
    }

    private func switchToAnonymousUser() {
        tracksService?.switchToAnonymousUser(withAnonymousID: NSUUID().uuidString)
    }

    private func switchToWordPressDotComUser() {
        tracksService?.switchToAuthenticatedUser(withUsername: "astralbodies", userID: "67137", skipAliasEventCreation: false)
    }
}

extension ViewController: NSFetchedResultsControllerDelegate {
    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChange anObject: Any, at indexPath: IndexPath?, for type: NSFetchedResultsChangeType, newIndexPath: IndexPath?) {
        self.updateObjectCountLabel()
    }
}
